### PR TITLE
chapter 3 solutions basic

### PR DESCRIPTION
**Pull request by pkk**
### Solutions for Chapter 3

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter3.hs [hunk 1]
```diff
@@ -374,6 +381,24 @@ after the fight. The battle has the following possible outcomes:
...
+  monsterHealth :: Int,
+  monsterAttack :: Int,
+  monsterGold :: Int
+} deriving (Show)
+
+fight :: Knight -> Monster -> Int
+fight knight monster
+  | knightAttack knight >= monsterHealth monster = knightGold knight + monsterGold monster
+  | knightHealth knight <= monsterAttack monster = (-1)
+  | otherwise = knightGold knight
```
Comment (chshersh) [inline src/Chapter3.hs:400]: Great solution! 👏🏻

File: src/Chapter3.hs [hunk 2]
```diff
@@ -460,6 +485,29 @@ Create a simple enumeration for the meal types (e.g. breakfast). The one who
...
+  | PostLunchBreakfast
+  | PostLunchLunch
+  | PostLunchDinner
+  | PreDinnerBreakfast
+  | PreDinnerLunch
+  | PreDinnerDinner
+  | Dinner
+  | PostDinnerBreakfast
+  | PostDinnerLunch
+  | PostDinnerDinner
```
Comment (chshersh) [inline src/Chapter3.hs:509]: Yummy! 😋 
Hilarious menu 😅  I wish I could live a day like this...

File: src/Chapter3.hs [hunk 3]
```diff
@@ -480,6 +528,49 @@ After defining the city, implement the following functions:
...
+        One -> 1
+        Two -> 2
+        Three -> 3
+        Four -> 4
+    in
+      if sum (map houseToInt (cityHouses city)) >= 10
+      then city { cityCastle = CastleWithWalls name}
+      else city
+  _ -> city
+
```
Comment (chshersh) [inline src/Chapter3.hs:573]: Perfect solution for the task with Castle and Walls 💎

File: src/Chapter3.hs [hunk 4]
```diff
@@ -560,22 +651,31 @@ introducing extra newtypes.
...
 
-calculatePlayerDefense :: Int -> Int -> Int
-calculatePlayerDefense armor dexterity = armor * dexterity
+calculatePlayerDefense :: Armor -> Dexterity -> Defense
+calculatePlayerDefense (Armor armor) (Dexterity dexterity) = Defense (armor * dexterity)
 
-calculatePlayerHit :: Int -> Int -> Int -> Int
-calculatePlayerHit damage defense health = health + defense - damage
+calculatePlayerHit :: Damage -> Defense -> Health -> Health
+calculatePlayerHit (Damage damage) (Defense defense) (Health health) = Health (health + defense - damage)
```
Comment (chshersh) [inline src/Chapter3.hs:678]: Nice work on newtypes! :medal_sports: Newtypes are one very important feature in Haskell :+1:
You can read more about the newtype pattern in our blog post: https://kowainik.github.io/posts/haskell-mini-patterns#newtype

File: src/Chapter3.hs [hunk 5]
```diff
@@ -910,6 +1021,22 @@ Implement instances of "Append" for the following types:
...
+newtype Gold = Gold Int
+
+instance Append Gold where
+  append :: Gold -> Gold -> Gold
+  append (Gold x) (Gold y) = Gold (x + y)
+
+instance Append [a] where
+  append :: [a] -> [a] ->[a] 
+  --append a b = a ++ b -- eta reduce
+  append = (++) -- @TODO : Find how it infix operator style function are implmented
```
Comment (chshersh) [inline src/Chapter3.hs:1033]: That's very simple actually 🙂 
Functions in Haskell can be used in two forms: prefix and infix.

By default, named functions are used in prefix form:

```haskell
append list1 list2
```

and operators by default are used in infix form:

```haskell
x + y
```

However, you can use named functions in the infix form by taking function name in backtics:

```haskell
list1 `append` list2
```

Similarly, you can use operators in prefix form by taking them in parentheses:

```haskell
(+) x y
```

That's why you can eta-reduce here:

```haskell
append a b = a ++ b
append a b = (++) a b  -- it's the same as above
append a = (++) a      -- eta-reduce one argument
append = (++)          -- eta-reduce second argument

File: src/Chapter3.hs [hunk 6]
```diff
@@ -971,6 +1098,22 @@ implement the following functions:
...
+
+isWeekend :: Weekday -> Bool
+isWeekend weekday = weekday == Sunday || weekday == Saturday
+
+-- Bounded gives the minBound ( Sunday in this case) and maxBound (Saturday)
+-- Enum gives us succ to move to next in enum
+nextDay :: Weekday -> Weekday
+nextDay weekday
+  | weekday == maxBound = Monday
+  | otherwise = succ weekday
```
Comment (chshersh) [inline src/Chapter3.hs:1111]: That is right!

This is also a bit more maintainable. If you decide, for example, to have the "American" system of week days in the data type constructors, you won't need to change this function at all! 🙂

**Review by chshersh — APPROVED** [review]
Amazing job on all the task in Chapter Three! 🔝 💯 

I see that the advanced task is not solved (and it's indeed the most challenging task in the whole course). Do you need any hints for it?

**Comment by pkk** [comment]
@chshersh 

> Amazing job on all the task in Chapter Three! 🔝 💯
> 
> I see that the advanced task is not solved (and it's indeed the most challenging task in the whole course). Do you need any hints for it?

Thanks. I go the advance task done but wasn't happy with it. I would probably redo chap 3 again before or after chap 4. 

BTW.. what is the use of eta reduce ? In some cases I find non eta reduced functions to be more readble.

e.g. function to find lenght of a list

`mylen = foldr (\_ acc -> acc + 1) 0 ` vs `mylen = foldr (\_ -> (+) 1) 0`

It could come down to personal choice but wanted to check if they are any limitations to use non eta reduced functions

**Comment by chshersh** [comment]
@pkk You're correct, that sometimes eta-reduced functions are harder to read vs non-eta-reduced functions. So there's no general rule when to eta-reduce or not, use your judgement 🙂 

But eta-reduced functions are helpful, when your functions have a lot of parameters and you want to partially apply some of them by default to provide a nicer API. For example:

```haskell
createAppWith :: Settings -> Env -> RunningMode -> Application
createAppWith ... = ... some implementation ...

-- using the default Settings
createAppDefault :: Env -> RunningMode -> Application
createAppDefault = createAppWith defaultSettings  -- see how we eta-reduce here and avoid passing Env and RunningMode arguments?
```

**Pull request merged**